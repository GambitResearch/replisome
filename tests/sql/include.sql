\set VERBOSITY terse
\pset format unaligned

-- predictability
SET synchronous_commit = on;

DROP TABLE IF EXISTS t1;
DROP TABLE IF EXISTS t2;
DROP TABLE IF EXISTS t3;

CREATE TABLE t1 (id int PRIMARY KEY);
CREATE TABLE t2 (id int PRIMARY KEY);
CREATE TABLE t3 (id int PRIMARY KEY);
CREATE TABLE s1 (id int PRIMARY KEY);

SELECT slot_create();


-- Include commands in single tables

insert into t1 values (1);
insert into t2 values (2);
insert into t3 values (3);
update t1 set id = 10 where id = 1;
update t2 set id = 20 where id = 2;
update t3 set id = 30 where id = 3;
delete from t1 where id = 10;
delete from t2 where id = 20;
delete from t3 where id = 30;

SELECT data FROM slot_get(
	'include', '{"table": "t1"}', 'include', '{"table": "t3"}');


-- Include commands on a pattern of tables

insert into t1 values (1);
insert into t2 values (2);
insert into s1 values (3);

SELECT data FROM slot_get(
	'include', '{"tables": "t"}');


insert into t1 values (4);
insert into t2 values (5);
insert into s1 values (6);

SELECT data FROM slot_get(
	'include', '{"tables": "^.1$"}');


-- Exclude a table after inclusion

insert into t1 values (7);
insert into t2 values (8);
insert into t3 values (9);
insert into s1 values (10);

SELECT data FROM slot_get(
	'include', '{"tables": "^t"}', 'exclude', '{"table": "t2"}');


-- Exclude a single table

insert into t1 values (11);
insert into t2 values (12);
insert into t3 values (13);


SELECT data FROM slot_get(
	'exclude', '{"table": "t2"}');


-- Exclude a pattern

insert into t1 values (14);
insert into t2 values (15);
insert into t3 values (16);
insert into s1 values (17);


SELECT data FROM slot_get(
	'exclude', '{"tables": ".1"}');


-- Include after exclusion

insert into t1 values (18);
insert into t2 values (19);
insert into t3 values (20);
insert into s1 values (21);


SELECT data FROM slot_get(
	'exclude', '{"tables": "t"}', 'include', '{"tables": ".2"}');


SELECT slot_drop();
