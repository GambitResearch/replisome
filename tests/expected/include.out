\set VERBOSITY terse
\pset format unaligned
-- predictability
SET synchronous_commit = on;
DROP TABLE IF EXISTS t1;
NOTICE:  table "t1" does not exist, skipping
DROP TABLE IF EXISTS t2;
NOTICE:  table "t2" does not exist, skipping
DROP TABLE IF EXISTS t3;
NOTICE:  table "t3" does not exist, skipping
CREATE TABLE t1 (id int PRIMARY KEY);
CREATE TABLE t2 (id int PRIMARY KEY);
CREATE TABLE t3 (id int PRIMARY KEY);
CREATE TABLE s1 (id int PRIMARY KEY);
SELECT slot_create();
slot_create
init
(1 row)
-- Include commands in single tables
insert into t1 values (1);
insert into t2 values (2);
insert into t3 values (3);
update t1 set id = 10 where id = 1;
update t2 set id = 20 where id = 2;
update t3 set id = 30 where id = 3;
delete from t1 where id = 10;
delete from t2 where id = 20;
delete from t3 where id = 30;
SELECT data FROM slot_get(
	'include', '{"table": "t1"}', 'include', '{"table": "t3"}');
data
{
	"change": [
		{
			"op": "I",
			"schema": "public",
			"table": "t1",
			"colnames": ["id"],
			"coltypes": ["int4"],
			"values": [1]
		}
	]
}
{
	"change": [
		{
			"op": "I",
			"schema": "public",
			"table": "t3",
			"colnames": ["id"],
			"coltypes": ["int4"],
			"values": [3]
		}
	]
}
{
	"change": [
		{
			"op": "U",
			"schema": "public",
			"table": "t1",
			"values": [10],
			"keynames": ["id"],
			"keytypes": ["int4"],
			"oldkey": [1]
		}
	]
}
{
	"change": [
		{
			"op": "U",
			"schema": "public",
			"table": "t3",
			"values": [30],
			"keynames": ["id"],
			"keytypes": ["int4"],
			"oldkey": [3]
		}
	]
}
{
	"change": [
		{
			"op": "D",
			"schema": "public",
			"table": "t1",
			"oldkey": [10]
		}
	]
}
{
	"change": [
		{
			"op": "D",
			"schema": "public",
			"table": "t3",
			"oldkey": [30]
		}
	]
}
(6 rows)
-- Include commands on a pattern of tables
insert into t1 values (1);
insert into t2 values (2);
insert into s1 values (3);
SELECT data FROM slot_get(
	'include', '{"tables": "t"}');
data
{
	"change": [
		{
			"op": "I",
			"schema": "public",
			"table": "t1",
			"colnames": ["id"],
			"coltypes": ["int4"],
			"values": [1]
		}
	]
}
{
	"change": [
		{
			"op": "I",
			"schema": "public",
			"table": "t2",
			"colnames": ["id"],
			"coltypes": ["int4"],
			"values": [2]
		}
	]
}
(2 rows)
insert into t1 values (4);
insert into t2 values (5);
insert into s1 values (6);
SELECT data FROM slot_get(
	'include', '{"tables": "^.1$"}');
data
{
	"change": [
		{
			"op": "I",
			"schema": "public",
			"table": "t1",
			"colnames": ["id"],
			"coltypes": ["int4"],
			"values": [4]
		}
	]
}
{
	"change": [
		{
			"op": "I",
			"schema": "public",
			"table": "s1",
			"colnames": ["id"],
			"coltypes": ["int4"],
			"values": [6]
		}
	]
}
(2 rows)
-- Exclude a table after inclusion
insert into t1 values (7);
insert into t2 values (8);
insert into t3 values (9);
insert into s1 values (10);
SELECT data FROM slot_get(
	'include', '{"tables": "^t"}', 'exclude', '{"table": "t2"}');
data
{
	"change": [
		{
			"op": "I",
			"schema": "public",
			"table": "t1",
			"colnames": ["id"],
			"coltypes": ["int4"],
			"values": [7]
		}
	]
}
{
	"change": [
		{
			"op": "I",
			"schema": "public",
			"table": "t3",
			"colnames": ["id"],
			"coltypes": ["int4"],
			"values": [9]
		}
	]
}
(2 rows)
-- Exclude a single table
insert into t1 values (11);
insert into t2 values (12);
insert into t3 values (13);
SELECT data FROM slot_get(
	'exclude', '{"table": "t2"}');
data
{
	"change": [
		{
			"op": "I",
			"schema": "public",
			"table": "t1",
			"colnames": ["id"],
			"coltypes": ["int4"],
			"values": [11]
		}
	]
}
{
	"change": [
		{
			"op": "I",
			"schema": "public",
			"table": "t3",
			"colnames": ["id"],
			"coltypes": ["int4"],
			"values": [13]
		}
	]
}
(2 rows)
-- Exclude a pattern
insert into t1 values (14);
insert into t2 values (15);
insert into t3 values (16);
insert into s1 values (17);
SELECT data FROM slot_get(
	'exclude', '{"tables": ".1"}');
data
{
	"change": [
		{
			"op": "I",
			"schema": "public",
			"table": "t2",
			"colnames": ["id"],
			"coltypes": ["int4"],
			"values": [15]
		}
	]
}
{
	"change": [
		{
			"op": "I",
			"schema": "public",
			"table": "t3",
			"colnames": ["id"],
			"coltypes": ["int4"],
			"values": [16]
		}
	]
}
(2 rows)
-- Include after exclusion
insert into t1 values (18);
insert into t2 values (19);
insert into t3 values (20);
insert into s1 values (21);
SELECT data FROM slot_get(
	'exclude', '{"tables": "t"}', 'include', '{"tables": ".2"}');
data
{
	"change": [
		{
			"op": "I",
			"schema": "public",
			"table": "t2",
			"colnames": ["id"],
			"coltypes": ["int4"],
			"values": [19]
		}
	]
}
{
	"change": [
		{
			"op": "I",
			"schema": "public",
			"table": "s1",
			"colnames": ["id"],
			"coltypes": ["int4"],
			"values": [21]
		}
	]
}
(2 rows)
-- Include specific column
CREATE TABLE tc1 (id int PRIMARY KEY, foo text, bar text);
CREATE TABLE tc2 (id int PRIMARY KEY, foo text, bar text);
-- Columns must be a list
SELECT data FROM slot_get(
	'include', '{"table": "tc1", "columns": 42}');
ERROR:  member "columns" must be a json array, in "{"table": "tc1", "columns": 42}"
SELECT data FROM slot_get(
	'include', '{"table": "tc1", "columns": ["id", "bar", "qux"]}');
data
(0 rows)
SELECT slot_drop();
slot_drop
stop
(1 row)
